File: lib/data/dto/leaderboard_dto.dart
class LeaderboardEntryDto {
  final int rank;
  final String userId;
  final String username;
  final String? avatarUrl;
  final int coins;

  const LeaderboardEntryDto({
    required this.rank,
    required this.userId,
    required this.username,
    this.avatarUrl,
    required this.coins,
  });

  factory LeaderboardEntryDto.fromJson(Map<String, dynamic> json) =>
      LeaderboardEntryDto(
        rank: (json['rank'] as num).toInt(),
        userId: json['userId'] as String,
        username: json['username'] as String,
        avatarUrl: json['avatarUrl'] as String?,
        coins: (json['coins'] as num).toInt(),
      );
}

class UserRankDto {
  final int rank;
  final int coins;
  final List<LeaderboardEntryDto> surrounding;

  const UserRankDto({
    required this.rank,
    required this.coins,
    required this.surrounding,
  });

  factory UserRankDto.fromJson(Map<String, dynamic> json) => UserRankDto(
        rank: (json['rank'] as num).toInt(),
        coins: (json['coins'] as num).toInt(),
        surrounding: (json['surrounding'] as List<dynamic>)
            .map((e) =>
                LeaderboardEntryDto.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

File: backend/src/leaderboard/dto/leaderboard.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { PaginationQueryDto, PaginationMeta } from '../../common/dto/pagination.dto';

export class LeaderboardEntryDto {
  @ApiProperty()
  rank: number;

  @ApiProperty()
  userId: string;

  @ApiProperty()
  username: string;

  @ApiPropertyOptional()
  avatarUrl?: string;

  @ApiProperty()
  coins: number;
}

export class LeaderboardQueryDto extends PaginationQueryDto {}

export class PaginatedLeaderboardDto {
  @ApiProperty({ type: [LeaderboardEntryDto] })
  data: LeaderboardEntryDto[];

  @ApiProperty()
  meta: PaginationMeta;
}

export class UserRankDto {
  @ApiProperty()
  rank: number;

  @ApiProperty()
  coins: number;

  @ApiProperty({ type: [LeaderboardEntryDto], description: 'Users around current user' })
  surrounding: LeaderboardEntryDto[];
}
